internal/proxy/proxy.go

```diff
@@ -188,95 +188,28 @@ func NewClient(ctx context.Context, cmd *cobra.Command, conf *Config) (*Client,
 		}
 	}
 
-	var mnts []*socketMount
 	for _, inst := range conf.Instances {
-		go func(name string) {
-			// Initiate refresh operation
-			d.EngineVersion(ctx, name)
-		}(inst.Name)
+		// Initiate refresh operation and warm the cache.
+		go func(name string) { d.EngineVersion(ctx, name) }(inst.Name)
 	}
+
+	var mnts []*socketMount
 	pc := newPortConfig(conf.Port)
 	for _, inst := range conf.Instances {
 		version, err := d.EngineVersion(ctx, inst.Name)
 		if err != nil {
 			return nil, err
 		}
 
-		var (
-			// network is one of "tcp" or "unix"
-			network string
-			// address is either a TCP host port, or a Unix socket
-			address string
-		)
-		// IF
-		//   a global Unix socket directory is NOT set AND
-		//   an instance-level Unix socket is NOT set
-		//   (e.g.,  I didn't set a Unix socket globally or for this instance)
-		// OR
-		//   an instance-level TCP address or port IS set
-		//   (e.g., I'm overriding any global settings to use TCP for this
-		//   instance)
-		// use a TCP listener.
-		// Otherwise, use a Unix socket.
-		if (conf.UnixSocket == "" && inst.UnixSocket == "") ||
-			(inst.Addr != "" || inst.Port != 0) {
-			network = "tcp"
-
-			a := conf.Addr
-			if inst.Addr != "" {
-				a = inst.Addr
-			}
-
-			var np int
-			switch {
-			case inst.Port != 0:
-				np = inst.Port
-			case conf.Port != 0:
-				np = pc.nextPort()
-			default:
-				np = pc.nextDBPort(version)
-			}
-
-			address = net.JoinHostPort(a, fmt.Sprint(np))
-		} else {
-			network = "unix"
-
-			dir := conf.UnixSocket
-			if dir == "" {
-				dir = inst.UnixSocket
-			}
-			if _, err := os.Stat(dir); err != nil {
-				return nil, err
-			}
-			address = UnixAddress(dir, inst.Name)
-			// When setting up a listener for Postgres, create address as a
-			// directory, and use the Postgres-specific socket name
-			// .s.PGSQL.5432.
-			if strings.HasPrefix(version, "POSTGRES") {
-				// Make the directory only if it hasn't already been created.
-				if _, err := os.Stat(address); err != nil {
-					if err = os.Mkdir(address, 0777); err != nil {
-						return nil, err
-					}
-				}
-				address = UnixAddress(address, ".s.PGSQL.5432")
-			}
-		}
-
-		var opts []cloudsqlconn.DialOption
-		if inst.IAMAuthN != nil {
-			opts = append(opts, cloudsqlconn.WithDialIAMAuthN(*inst.IAMAuthN))
-		}
-		m := &socketMount{inst: inst.Name, dialOpts: opts}
-		addr, err := m.listen(ctx, network, address)
+		m, err := newSocketMount(ctx, conf, pc, inst, version)
 		if err != nil {
 			for _, m := range mnts {
-				m.close()
+				_ = m.Close() // TODO: log this error?
```

Comment:
Seems like we should be consistent and log the error here if we are logging it when the Client closes